[8.0] add dirac-login user guide, code docs (#5397)
DIRACGrid/DIRAC: docs/source/UserGuide/GettingStarted/GettingUserIdentity/index.rst

  $ dirac-proxy-init -g <user_group>

where ``user_group`` is the desired DIRAC group name for which the user is entitled.

> .. note:: Version 8.0 added the possibility to generate proxy with new `dirac-login` command, use *--help* switch for more information. E.g.: dirac-login <user_group>

Author: fixed

Token authorization
-------------------

Starting with the 8.0 version of DIRAC, it is possible to authorize users through third party Identity Providers (IdP),

> such as EGI Checkin [https://www.egi.eu/services/check-in/] or WLCG IAM (https://indigo-iam.github.io/v/current/).

Author: fixed

You do not need a certificate for this in a terminal, the main thing is that you must be registered in one of the supported IdP. The registration process is different for each IdP.

Once your account is created, you will be able to register with DIRAC Authorization Server using *--use-diracas* switch of the `dirac-login` command::

  dirac-login <user_group> --use-diracas

You can request to return the access token instead of a proxy using *--token* key::

  dirac-login <user_group> --token

But since not all services currently support tokens, you can get a proxy if you use the *--proxy* key::

  dirac-login <user_group> --proxy --use-diracas

> .. note:: if you want to get a proxy after logging in to DIRAC Authorization Server you must first put it in DIRAC, see "Proxy initialization".

Author: Yes, there is no magic here. As we have agreed, DIRAC will not issue **self-signed** proxies or tokens, so DIRAC must receive them first in order to be able to give them to a user or service (to perform asynchronous tasks on behalf of the user).

**1.**
There are several ways to "upload" a proxy to DIRAC, you know them (+ `dirac-login`). This thing has not changed.
To avoid this, I consider `RCAuth` (in progress), which generates a certificate and a proxy on the fly, but it has significant **limitations**.

**2.**
After that, you can make
```
dirac-login dirac-user --use-diracas
```
 from any installation (not server) without a certificate and after authorization through IAM for example you will receive a proxy.

Or you can configure a client without a certificate (without `dirac-proxy-init`) using `--login` flag (and for this action not need proxy in `ProxyManager`, it can be done with tokens because all involved components running with tornado):
```
dirac-configure --login=https://lbcertifdiracoauth.cern.ch/auth
```

> .. note:: if you want to get a proxy after logging in to DIRAC Authorization Server you must first put it in DIRAC, see "Proxy initialization".

Reviewer: For my understanding: is this a workflow that we can consider "advised" or "suggested"? I guess this comes from part of the discussions about proxies -> tokens transition.

> .. note:: if you want to get a proxy after logging in to DIRAC Authorization Server you must first put it in DIRAC, see "Proxy initialization".

Author: Sorry for the late answer, as far as I understand the situation around the discussions about the use of tokens and how they are already integrated into other products, I would say that we recommend trying to use tokens, as the transition to tokens is expected in the near future.

I think that in most cases users will prefer to use:
```
dirac-login my-group
```
or `dirac-proxy-init` of course.

> .. note:: if you want to get a proxy after logging in to DIRAC Authorization Server you must first put it in DIRAC, see "Proxy initialization".

Author: When interacting with a user, it is important for DIRAC to obtain his **proxy** certificate and **refresh token**. Having received them once, it is expected that they should be enough for about a year. But if there are no problems with receiving tokens: you just need to have a browser and an account in IAM for example, then with a proxy a little harder: you need to have a local certificate to generate proxy.

  $ dirac-proxy-init -g <user_group>

where ``user_group`` is the desired DIRAC group name for which the user is entitled.

.. versionadded:: 8.0
   added the possibility to generate proxy with new `dirac-login` command, use *--help* switch for more information. E.g.: dirac-login <user_group>

Token authorization
-------------------

Starting with the 8.0 version of DIRAC, it is possible to authorize users through third party Identity Providers (IdP),
such as `EGI Checkin <https://www.egi.eu/services/check-in/>`_ or `WLCG IAM <https://indigo-iam.github.io/v/current/>`_.

> You do not need a certificate for this in a terminal, the main thing is that you must be registered in one of the supported IdP. The registration process is different for each IdP.

Reviewer: ```suggestion
You do not need a certificate for this in a terminal, but you must be registered in one of the supported IdP. The registration process is different for each IdP.
```

  $ dirac-proxy-init -g <user_group>

where ``user_group`` is the desired DIRAC group name for which the user is entitled.

Token authorization
-------------------

> Starting with the v7r4 version in DIRAC, it is possible to authorize users through third party Identity Providers (IdP),

Reviewer: ```suggestion
Starting with the 8.0 version of DIRAC, it is possible to authorize users through third party Identity Providers (IdP),
```

> such as EGI Checkin [https://www.egi.eu/services/check-in/] or WLCG IAM (https://indico.cern.ch/event/739896/contributions/3497694/attachments/1905332/3146590/IAM-WLCG-AuthZ-Fermilab-10092019.pdf).

Reviewer: Please use proper rst links. 
For IaM maybe https://github.com/indigo-iam/iam or https://indigo-iam.github.io/v/current/ are better links

To do this, you do not need to have a certificate if you use a terminal, the main thing is that you must be registered in one of the supported IdP. The registration process is different for each IdP.

Once your account is created, you will be able to register with DIRAC using the `dirac-login` command that will return tokens that will be used to access the services::

  dirac-login -g <user_group>

But since not all services currently support tokens, you can get a proxy if you use the "--proxy" key::

>   dirac-login -g <user_group> --proxy

Author: the result is the same - a proxy with a group.

I can add a "--use-certificate" key to generate a proxy completely on the client side (as dirac-proxy-init does).

dirac-proxy-init analog:

```
dirac-login --use-certificate
```

Login without local certificat to get proxy:

```
dirac-login --proxy
```

To get token:

```
dirac-login
```

What do you think?

>   dirac-login -g <user_group> --proxy

Reviewer: Yes, I think it would make sense -- one command to rule them all.

>   dirac-login -g <user_group> --proxy

Reviewer 2: Can this command be made smarter so users don't need to understand the weird transitional mess?

* Perhaps `--proxy` should be the default for now and `--no-proxy` can be used instead if you really don't want one.
* `--use-certificate` is enabled by default if one is available (`--use-certificate`/`--no-use-certificate` can be used to make it explicit)

>   dirac-login -g <user_group> --proxy

Author: In order to avoid clutter when starting to use tokens, I left `dirac-proxy-init` (which everyone knows well) and added a new `dirac-login` method, which is fundamentally differs in the method of authorization (via authorization server).

Some notes:

- At authorization we can receive tokens or proxies.
- I think we can assume that very soon all services will be tornadoed and DIRAC will still have something to use.
- OAuth2 framework uses tokens, if you remember our discussions and agree to do everything as standard as possible, then I think the tokens should be the default for dirac-login.
- on the client side there may be requests to non-DIRAC resources with a proxy or token (received through DIRAC AS).

IMHO:
> Perhaps --proxy should be the default for now and --no-proxy can be used instead if you really don't want one.

I should be --proxy and --tokens.

I'll think about what the script can decide without asking the user, so as not to be too clever.

>   dirac-login -g <user_group> --proxy

Author: I did not find the perfect solution, but given the wishes did the following:

Choose authorization way: `--use-certificate`/`--use-diracas`
Choose response type: `--proxy`/`--token`
By default: `--use-certificate` `--proxy`

1. create proxy locally with certificate:
```
dirac-login
# will return proxy
```

2. Login through DIRAC Authorization Server:
```
dirac-login --use-diracas
# will return proxy
``` 
```
dirac-login --token
# will return token
```
